web-vr/hello-world/demo/demo-vr.js

```diff
@@ -0,0 +1,203 @@
+/*!
+ *
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+/* eslint-env es6 */
+
+class DemoVR extends Demo {
+  constructor () {
+    super();
+
+    this._disabled = false;
+    if (typeof VRFrameData === 'undefined') {
+      this._disabled = true;
+      this._showError();
+      return;
+    }
+
+    this._button = undefined;
+    this._vr = {
+      display: null,
+      frameData: new VRFrameData()
+    };
+
+    this._addVREventListeners();
+    this._getDisplays();
+  }
+
+  _addVREventListeners () {
+    window.addEventListener('vrdisplayactivate', evt => {
+      evt.preventDefault();
+      this._activateVR();
+    });
+
+    window.addEventListener('vrdisplaydeactivate', evt => {
+      evt.preventDefault();
+      this._deactivateVR();
+    });
+  }
+
+  _getDisplays () {
+    return navigator.getVRDisplays().then(displays => {
+      if (displays.length === 0) {
+        return;
+      }
+
+      this._vr.display = displays[0];
+      this._vr.depthNear = DemoVR.CAMERA_SETTINGS.near;
+      this._vr.depthFar = DemoVR.CAMERA_SETTINGS.far;
+
+      if (!this._vr.display.capabilities.canPresent) {
+        return this._showNoPresent();
+      }
+
+      this._createPresentationButton();
+    });
+  }
+
+  _showNoPresent () {
+    console.warn(`Unable to present with this device ${this._vr.display}`);
+  }
+
+  _showError () {
+    console.error('WebVR not supported');
+  }
+
+  _createPresentationButton () {
+    this._button = document.createElement('button');
+    this._button.classList.add('vr-toggle');
+    this._button.addEventListener('click', _ => {
+      this._toggleVR();
+    });
+    document.body.appendChild(this._button);
+
+    this._updateButtonLabel();
+  }
+
+  _updateButtonLabel () {
+    this._button.textContent = this._vr.display.isPresenting ?
+        'Disable VR' :
+        'Enable VR';
+  }
+
+  _deactivateVR () {
+    if (!this._vr.display) {
+      return;
+    }
+
+    this._vr.display.exitPresent();
```

Comment:
1. You return a `Promise` from this method, but I don't know why
1. `activateVR` does NOT return a `Promise`, even though it's similar
1. `exitPresent` and `requestPresent` return `Promise`s, if you really want to return one.
1. Caveat - `exitPresent` will reject if you weren't actually presenting